docs/content/sips/007-deployment-auth.md

```diff
@@ -0,0 +1,158 @@
+* `--bindle-password` or `$BINDLE_PASSWORD` - optional
+* `--hippo-server` or `$HIPPO_URL` - required
+* `--hippo-username` or `$HIPPO_USERNAME` - required
+* `--hippo-password` or `$HIPPO_PASSWORD` - required
+
+## Proposal
+
+### Login command
+
+We propose to add a `spin login` command, which takes the desired Hippo server as an argument:
+
+`spin login --url https://example.com`
+
+This will perform the GitHub device authorisation flow.  The outline user experience (we'll dive into more detail below) is:
+
+```
+$ spin login --url https://example.com
+Open the Platform device authorization URL in your browser: https://example.com/device-authorization and enter the code: tWOJCFhY
+Waiting for device authorization...
+Waiting for device authorization...
+Device authorized!
+Login successful.
+```
+
+The URL and token generated by the device authorisation flow is stored on the machine.
+
+A user may log into multiple Hippo servers.  The (URL, token) pairs for each will be stored.  That is, logging into one server does not log you out of others.
+
+> QUESTION: Or should you only be logged into one at a time?
+
+> QUESTION: Should we provide an explicit `logout` command to destroy the token?
+
+### Deploy command changes
+
+The `spin deploy` command will now act as follows:
+
+* The Hippo URL, username and password are now **optional** at the implementation level.
+* If the Hippo URL is specified:
+  * If the user is logged into that server, use the stored token for auth.  The username and password are not required and _must not be present_.
+  * If the user is not logged into that server, and a username or password is provided, follow the current behaviour.  Both username and password are required.
+  * If the user is not logged into that server, and no username or password is provided, attempt a login using the same behaviour as `spin login`.  If that succeeds, continue to deployment.  If it fails, exit.
+* If the Hippo URL is **not** specified:
+  * If a username or password is specified, error. A URL is required in this case.
+  * If the user is logged in, use the stored URL and token.
+  * If the user is **not** logged in, attempt to log them into a default URL, such as the default address of a local Fermyon platform. If that fails, print usage information and exit.
+
+> QUESTION: If a user is logged into multiple servers, should we:
+> * Default the URL to the _most recent_ login
+> * Provide a way for the user to say which URL is their default
+> * Require the user to provide a URL
+> * Provide an abstraction (e.g. name, config file path) by which the user can switch between servers
+
+> QUESTION: The last section represents a change in behaviour when no flags are provided. Are we happy with this, or should we direct people to `spin login` and not attempt it automatically?
+
+> QUESTION: Is it correct to reject a username/password if the user is logged in to the chosen server, or should we attempt username/password login in this case?
```

Comment:
@bacongobbler Can you clarify the comment "The first time the user runs spin deploy, they will be logged in" please?  It sounds like this is postulating:

1. There will be a concept of "logging into" a username-password server (as well as a device token server)
2. This will happen _implicitly_ if the username-password set of options are provided during `deploy` (rather than controlling the current deployment but not affecting state).  [Token login is an explicit step, even if triggered from `deploy`, and is currently specced not to happen if username-password options are present.]

I am not keen on 2, though we could do it if we ask the user "do you want to remain logged in" or some such wording.

But I agree, if 1 is true then we need to rework this to express that if URL/username/password are present and the URL is _not_ logged in as a token server, then URL/username/password should override and behave as they do now.

Reply:
I think if credentials are explicitly set via CLI arg or env var, stored credentials should be ignored. It could make sense to include the username with stored credentials and warn a user if they are overriding that saved username's credential, i.e. if I login as `lann` and later `HIPPO_USERNAME=lann HIPPO_PASSWORD=... spin deploy`, I might get a message `You are setting a password for already logged-in user "lann".` to help contextualize any auth errors.

I would expect credentials to only be saved if 1) they are entered interactively or 2) they are explicitly passed to `spin login`.

Reply:
I had considered the point of view where `spin deploy` does not accept "login" feature flags at all. That is to say, `HIPPO_USERNAME`, `HIPPO_PASSWORD`, etc. are all localized to `spin login`.

Reply:
Sorry, had to jump in a meeting so I had to leave mid-thought.

> There will be a concept of "logging into" a username-password server (as well as a device token server)

Yes.

> This will happen implicitly if the username-password set of options are provided during `deploy`

Not entirely. I expect the user to run `spin login` prior to running `spin deploy`. All login-related feature flags like `--hippo-username` will be removed from `spin deploy` and moved to `spin login`.

My mental model looks like this:

1. An explicit login step (via `spin login`) is required. Envvars like `HIPPO_USERNAME` are read during `spin login`.
2. Follow-up commands that require login session information (like `spin deploy`) will read session information that is stored in the keyring.
3. Commands like `spin deploy` do NOT have an option to pass a username/password/login URL.

This mental model maps closely with `docker`:

1. `docker login`: log into the server via username/password, storing session information into the keyring
2. `docker push`: pull session information from the keyring; push docker image to the server
3. `docker pull`: pull session information from the keyring; pull docker image from the server
4. `docker logout`: remove session information from the keyring.

This keeps commands like `spin deploy` focused on feature flags related to deploying the application. Login-related feature flags are strictly handled by `spin login` and ONLY `spin login`.

Here's what `docker push --help` looks like:

```
><> docker push --help

Usage:  docker push [OPTIONS] NAME[:TAG]

Push an image or a repository to a registry

Options:
  -a, --all-tags                Push all tagged images in the repository
      --disable-content-trust   Skip image signing (default true)
  -q, --quiet                   Suppress verbose output
```

Notice how there's nothing related to logging into the registry,

Reply:
Let's say I have a CI/CD pipeline that wants to deploy a staging app, with a auth token in some external secret management like a Github actions secret. I don't _really_ want that secret persisted anywhere on disk. Ideally I could do something like `export HIPPO_TOKEN=...` at the beginning of the script and expect it to just work.

Reply:
We could make an exception for accepting `HIPPO_TOKEN` as input to `spin deploy`. We've talked about supporting personal access tokens for CI/CD pipeline scenarios where the login flow doesn't make much sense.

That being said, I _really_ want to keep the separation of concerns between "logging in" and "deploying an app" as separate as possible. Accepting a one-off auth token would be OK, but supporting the entire login workflow will be difficult to maintain over time. We convoluted the concerns of "logging in" and "pushing to a server" with [Draft](https://github.com/Azure/draft-classic) and it was an absolute nightmare to maintain as we introduced new features for `draft up`.